### PR TITLE
Add kubelet `SeccompDefault` alpha feature

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This adds the gate `SeccompDefault` as new alpha feature. Seccomp path and field fallbacks are now passed to the helper functions, whereas unit tests covering those code paths have been added as well.

#### Which issue(s) this PR fixes:

Refers to https://github.com/kubernetes/enhancements/issues/2413

#### Special notes for your reviewer:

The functions `fieldProfile` and `fieldSeccompProfile` usually never receive scmp `smp *v1.SeccompProfile == nil`. I decided to not cleanup those paths for security reasons.

Beside enabling the feature gate, the feature has to be enabled by the `SeccompDefault` kubelet configuration or its corresponding `--seccomp-default` CLI flag.

Another thing to mention is that the feature does not change the API at all. This means that people using the feature will not get a modified `SecurityContext` if they do not specify anything within it.

#### Does this PR introduce a user-facing change?

```release-note
Added new kubelet alpha feature `SeccompDefault`. This feature enables falling back to
the `RuntimeDefault` (former `runtime/default`) seccomp profile if nothing else is specified
in the pod/container `SecurityContext` or the pod annotation level. To use the feature, enable
the feature gate as well as set the kubelet configuration option `SeccompDefault`
(`--seccomp-default`) to `true`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/tree/34fa3dd/keps/sig-node/2413-seccomp-by-default
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
@mrunalp @pjbgf @liggitt PTAL

**Comment by liggitt** [comment]
looking at the PRR, it is unclear to me how an operator would detect that enabling this caused an issue

> How can a rollout fail? Can it impact already running workloads?
> 
> * Workloads on a node may starting to fail when (re)scheduled on the node where the feature is enabled. Required specific syscalls may be blocked by the default seccomp profile, which will cause the application to get terminated.

> What specific metrics should inform a rollback?
> 
> * If a workload is starting to fail because of blocked syscalls (audit logs), then a temporarily rollback would be appropriate in production.

can you clarify how a cluster operator would know if this was causing problems?

**Comment by liggitt** [comment]
> Pods which previously run successfully would now get some syscalls blocked they either need for startup of the workload or a single feature. This means that in the worst case the pod gets stuck in a crash loop or encounter degraded functionality. So a cluster operator would see restarting containers, while the logs would indicate something like "operation not permitted".

Detecting when a crash is due to this requires application-specific knowledge, right? I don't see a way for a cluster operator to detect this generically.

> A good pre-check would be to run an e2e test suite for the application with the runtime default profile if the application was running unconfined before.

That is a good idea for the application owner, but the application owner and cluster operator are often two different entities.

**Comment by saschagrunert** [comment]
> > Pods which previously run successfully would now get some syscalls blocked they either need for startup of the workload or a single feature. This means that in the worst case the pod gets stuck in a crash loop or encounter degraded functionality. So a cluster operator would see restarting containers, while the logs would indicate something like "operation not permitted".
> 
> Detecting when a crash is due to this requires application-specific knowledge, right? I don't see a way for a cluster operator to detect this generically.

The only way to get a hint to it is when toggling the feature toggles crashing pods within the cluster.

> > A good pre-check would be to run an e2e test suite for the application with the runtime default profile if the application was running unconfined before.
> 
> That is a good idea for the application owner, but the application owner and cluster operator are often two different entities.

Agreed, mixed up the roles in my mind.

**Comment by liggitt** [comment]
don't we need a plausible rollout story for this feature before starting implementation?

**Comment by mrunalp** [comment]
> don't we need a plausible rollout story for this feature before starting implementation?

Since this is targeting alpha, one approach would be for users to enable this on a staging server to ensure their applications and/or platform works as expected. If there are any changes needed such as setting it to unconfined or making custom policies for certain pods, they can test them on staging and once things look good they can roll this out to production (if they ever roll out alpha feature flags there  :) ).

**Comment by derekwaynecarr** [comment]
@liggitt @saschagrunert @mrunalp isnt this change rolled out with a corresponding update in container runtime configuration, as a result, the staging approach seems valid for this while its in alpha imo.  maybe we can update the prr for the alpha->beta transition in 1.23 based on what we observe in our own real world testing.

File: pkg/kubelet/kuberuntime/security_context.go
Comment (mrunalp) [reply pkg/kubelet/kuberuntime/security_context.go:39]: An option may be to add a kubelet flag that allows one to pick runtime/default, unconfined or custom. A similar idea was captured as an alternative in the KEP.

Comment (liggitt) [reply pkg/kubelet/kuberuntime/security_context.go:39]: Feature gates are intended to cover the development life cycle of a feature - they are not intended to be long-term APIs. If we expect admins to have kubelet-level control over this once this feature graduates to GA, this needs a dedicated kubelet configuration option

Comment (mrunalp) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > Feature gates are intended to cover the development life cycle of a feature - they are not intended to be long-term APIs.

Agree.

> If we expect admins to have kubelet-level control over this once this feature graduates to GA, this needs a dedicated kubelet configuration option.

Yes. I brought up the alternative if there are concerns that there is no opt out at a global level in the kubelet once this goes GA. 
I still think that it is fine if this is enabled by default when we go to GA. The users opt out of seccomp at the pod level as needed. 

cc: @tallclair

Comment (liggitt) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > I still think that it is fine if this is enabled by default when we go to GA. The users opt out of seccomp at the pod level as needed.

How do users on a managed cluster get that signal other than noticing they are broken on an upgrade to a version where this is enabled?

Comment (tallclair) [reply pkg/kubelet/kuberuntime/security_context.go:39]: Note that both cri-o and containerd let you configure the runtime default. IMO the configuration of the default should be left to the runtime, and Kubelet should (once this goes to GA) always defer to the runtime for the default, rather than making this configurable in the Kubelet.

**Comment by tallclair** [comment]
re: rollout - I don't think you need application specific knowledge to know if seccomp is causing a failure. It should show up in the kernel audit logs. Granted understanding those logs requires some specialized knowledge, but at least the presence of `type=SECCOMP` errors should signal something is up.

File: pkg/kubelet/kuberuntime/security_context.go
Comment (mrunalp) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > How do users on a managed cluster get that signal other than noticing they are broken on an upgrade to a version where this is enabled?

Thinking out aloud on how a managed service would roll out this out if there was a kubelet configuration option for this: 
Would they ask the users to opt-in to this feature as a cluster option?
If yes, how would the user know if they can safely do it?
They would have to first test their pods by enabling runtime/default on them (on older unupgraded cluster) and see if they work or set them to unconfined explicitly or use some custom policy for their workloads. I don't see how they can get by without testing and looking at logs.
If they are comfortable with this test, they will then go ahead with the update removing the runtime/default from the pod and rely on the cluster configuration for runtime/default instead.
If we go with this model, we will have some users that never opt-in to this feature because they don't want to test it or don't have the time to test till a CVE motivates them to do it.
This approach works, but doesn't improve the security by default. 

The approach in this PR is taking us on a path to forcing everyone to use seccomp by default.

So, we need to pick whether we want to force everyone to use seccomp by default or allow them to not use it as long as they want to. 

Even with force approach, someone can still craft a runtime default seccomp policy which is close to unconfined.

Comment (liggitt) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > The approach in this PR is taking us on a path to forcing everyone to use seccomp by default.

it seems like it's changing the meaning of an empty seccomp field in the Pod API... how is that not a breaking change?

Comment (mrunalp) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > it seems like it's changing the meaning of an empty seccomp field in the Pod API... how is that not a breaking change?

Yes, it is. So, what would be the preferred way to introduce this? Is the kubelet configuration option acceptable or something else?

Comment (liggitt) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > Is the kubelet configuration option acceptable or something else?

That's what I expected.

Comment (mrunalp) [reply pkg/kubelet/kuberuntime/security_context.go:39]: > That's what I expected.

Sounds good to me. @tallclair @saschagrunert @derekwaynecarr wdyt?

Comment (saschagrunert) [reply pkg/kubelet/kuberuntime/security_context.go:39]: I'm fine with that, too. As a summary what we now do:

- Replace the feature gate with a kubelet configuration option (CLI and config file)
- Update the KEP to reflect that change
- Reconsider alpha to beta upgrades within the KEP

Is this correct?

Comment (johnbelamaric) [reply pkg/kubelet/kuberuntime/security_context.go:39]: I don't think it's a good idea to release a feature that will break workloads on upgrade. We should require a flag to enable this even in GA.

Comment (saschagrunert) [reply pkg/kubelet/kuberuntime/security_context.go]: Added a `SeccompDefault` kubelet configuration as well as a corresponding CLI flag `--seccomp-default`. Both are now covered by the feature gate check and everything is disabled by default.

File: cmd/kubelet/app/options/options.go
Comment (pjbgf) [inline cmd/kubelet/app/options/options.go:168]: nit:
```suggestion
	// SeccompDefault enables the use of `RuntimeDefault` as the default seccomp profile for all workloads on the node.
```

File: cmd/kubelet/app/options/options.go
Comment (pjbgf) [inline cmd/kubelet/app/options/options.go:355]: nit:
```suggestion
	fs.BoolVar(&f.SeccompDefault, "seccomp-default", f.SeccompDefault, "<Warning: Alpha feature> Enable the use of `RuntimeDefault` as the default seccomp profile for all workloads. The SeccompDefault feature gate must be enabled to allow this flag, which is disabled per default.")
```

File: pkg/features/kube_features.go
Comment (pjbgf) [inline pkg/features/kube_features.go:740]: nit:
```suggestion
	// Enables the use of `RuntimeDefault` as the default seccomp profile for all workloads.
```

File: pkg/kubelet/apis/config/types.go
Comment (pjbgf) [inline pkg/kubelet/apis/config/types.go:410]: nit:
```suggestion
	// SeccompDefault enables the use of `RuntimeDefault` as the default seccomp profile for all workloads.
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (pjbgf) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:145]: nit:
```suggestion
	// Use RuntimeDefault as the default seccomp profile for all workloads.
```

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (pjbgf) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:867]: nit:
```suggestion
	// SeccompDefault enables the use of `RuntimeDefault` as the default seccomp profile for all workloads.
```

**Comment by saschagrunert** [comment]
Thanks for the review @pjbgf, I addressed all of your suggestions.

**Comment by pjbgf** [comment]
> re: rollout - I don't think you need application specific knowledge to know if seccomp is causing a failure. It should show up in the kernel audit logs. Granted understanding those logs requires some specialized knowledge, but at least the presence of `type=SECCOMP` errors should signal something is up.

Not sure whether it is worth mentioning, but correlating audit logs to pods and the failed syscalls may be a bit painful. It may be worth mentioning that [security-profiles-operator](https://github.com/kubernetes-sigs/security-profiles-operator) has a kernel audit logs enrichment feature that would allow for users to easily spot the container/pod and the respective blocked syscall:
 
image

**Comment by mikedanese** [comment]
Ref #39845, #81115

**Comment by saschagrunert** [comment]
Had to rebase again, this is still ready for review.

**Comment by mmdriley** [comment]
> don't we need a plausible rollout story for this feature before starting implementation?

> Detecting when a crash is due to this requires application-specific knowledge, right? I don't see a way for a cluster operator to detect this generically.

I want to strongly echo both these points. I have maintained a syscall sandbox for a production service, where the sandbox was maintained apart from the application itself. Bugs routinely manifested as terrifyingly slight differences in behavior that might not even result in crashes, much less make enough noise to get triaged.

If the bugs _do_ get triaged, debugging them is very difficult because they result in paths through the program that are "impossible" as far as the application code is concerned. Consider that `containerd`'s default profile chooses to [return errors for failed syscalls](https://github.com/containerd/containerd/blob/c8cbf7998b4dd871d30026317cb906bf17147513/contrib/seccomp/seccomp_default.go#L456-L460), adding entirely new error cases to applications that never needed to deal with them before.

Mention was made that these errors appear in the kernel audit logs. Maybe so. But how would an application developer or infrastructure engineer know to look there in the first place?

On GKE we are evaluating whether rolling this out to customer nodes would constitute a breaking change, and currently we're leaning toward "yes". I expect other infrastructure providers have similar concerns. It would be helpful to see a rollout plan with appropriate risk mitigations in place to avoid breaking workloads on node upgrade.

**Comment by mmdriley** [comment]
Ah, but I see @johnbelamaric already covered this in a comment:
> I don't think it's a good idea to release a feature that will break workloads on upgrade. We should require a flag to enable this even in GA.

Does that reflect the current plan, and is it expected the KEP will be updated?

**Comment by saschagrunert** [comment]
> Ah, but I see @johnbelamaric already covered this in a comment:
> 
> > I don't think it's a good idea to release a feature that will break workloads on upgrade. We should require a flag to enable this even in GA.
> 
> Does that reflect the current plan, and is it expected the KEP will be updated?

We definitely have to update the KEP reflecting the flag. I think together with that I can also add a rollout plan. :slightly_smiling_face:

**Comment by saschagrunert** [comment]
KEP update PR in https://github.com/kubernetes/enhancements/pull/2773

**Comment by saschagrunert** [comment]
The KEP update PR got merged, which means this is now ready for review. @liggitt @kubernetes/sig-node-pr-reviews please take a look again

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:867]: note this required the SeccompDefault feature gate?

**Comment by liggitt** [comment]
config API and cli bits lgtm (one nit on documenting the feature gate requirement)

**Comment by liggitt** [comment]
tag me once the functional bits of this have approval and I can add approve for the config API

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (saschagrunert) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:867]: Yep, added a comment as suggested :+1:

**Comment by saschagrunert** [comment]
Pinging for approval:
- cmd/kubelet/OWNERS: @mrunalp @sjenning 
- pkg/features/OWNERS: @dchen1107 @derekwaynecarr
- pkg/kubelet/OWNERS: @mrunalp @sjenning

**Review by mrunalp — APPROVED** [review]

**Comment by liggitt** [comment]
/approve

based on @mrunalp lgtm (feature gate approved by @derekwaynecarr in the KEP)

**Pull request merged**
